libminifi/include/utils/StringUtils.h

```diff
@@ -137,6 +137,7 @@ class StringUtils {
 
   static std::vector<std::string> split(const std::string &str, const std::string &delimiter);
   static std::vector<std::string> splitAndTrim(const std::string &str, const std::string &delimiter);
+  static std::vector<std::string> splitAndTrimOnString(std::string str, const std::string& delimiter);
```

Comment:
How is this different from splitAndTrim?

Reply:
`splitAndTrim` only uses the first character of the given string as a delimiter, so it actually is using a character as a delimiter not a string. I'm not sure if that should be updated to have a character as an input delimiter and keep 2 overloads or the functions should be merged.

Reply:
3 reasonable ways of using the delimiter string if it's longer than a character is
1. Match any of the characters in the string. Strings like this would be reasonable: `" \t;:,"`. C# does this.
2. Match the whole delimiter string as a delimiter. This is the case with ...OnString now. JavaScript and Python does this.
3. Match the whole delimiter string as a regex. Java does this, but I wouln't go this route.

Taking a string and only using its first character is a bit strange to me, I would change that behavior. The fact that it's trimmed after splitting should be relatively separate, but convenience functions are OK.

I can't think of a good and simple solution that does it all and that I wouldn't want to change eventually, so I'm not making any suggestions, but I'm happy to provide my opinion if you want to change it. Please consider adding comments that clarify the behavior.

Reply:
wow this is "great", I wondered why the following test passed knowing this:
```
TEST_CASE("TestStringUtils::split4", "[test split classname]") {
  std::vector<std::string> expected = { "org", "apache", "nifi", "minifi", "utils", "StringUtils" };
  REQUIRE(expected == StringUtils::split(org::apache::nifi::minifi::core::getClassName<org::apache::nifi::minifi::utils::StringUtils>(), "::"));
}
```
turns out two wrongs make a right here :D 
the current implementation implicitly discards zero length segments, so `split("a,,b", ",") == ["a", "b"]` instead of `["a", "", "b"]`, I would definitely change this, I would keep a single function with `std::string` delimiter but with the "correct" semantics, if we would like to interpret each character as a delimiter later, we could add a `splitAny` or something along those lines

Reply:
Thanks for both of your inputs. I changed `split` and `splitAndTrim` to use full strings and changed the semantics to keep the zero length segments. I also added `*RemovingEmpty` versions for utility. Changes are incorporated in 3ee4f1357bf03ec5bed69a3ca63801d128e739f8

Reply:
Because of the new semantics of `split` there are still some issues that need to be resolved, putting it to draft until it's fixed.

Reply:
Fixed the remaining issues.